File: goblingardenguards/goblingardenguards.cpp
#include <iostream>
#include <string>
#include <set>
#include <algorithm>
#include <cmath>

using namespace std;

class Goblin {
	public:
		int x;
		int y;
		int count;

		Goblin(int x, int y, int count = 1) {
			this->x = x;
			this->y = y;
			this->count = count;
		}

		bool operator<(const Goblin& rhs) const {
			return this->x * 10001 + this->y < rhs.x * 10001 + rhs.y;
		}

		bool operator>(const Goblin& rhs) const {
			return rhs < *this;
		}
};

ostream& operator<<(ostream& os, const Goblin& g) {
	return os << "(" << g.x << ", " << g.y << "): " << g.count;
}

bool in_range(const Goblin& g, int x, int y, int r) {
	return sqrt((g.x - x)*(g.x - x) + (g.y - y)*(g.y - y)) <= r;
}

int main() {
	int nbr_goblins;
	cin >> nbr_goblins;
	set<Goblin> goblins;

	for (int i = 0; i < nbr_goblins; i++) {
		int x, y;
		cin >> x >> y;
		Goblin g(x,y);
		pair<set<Goblin>::iterator, bool> res = goblins.insert(g);
		if (!res.second) {
			g.count += res.first->count;
			goblins.erase(res.first);
			goblins.insert(++res.first, g);
		}
	}

	int count = 0;
	int nbr_sprinklers;
	cin >> nbr_sprinklers;
	for (int i = 0; i < nbr_sprinklers; i++) {
		int x, y, r, xmin, xmax, ymin, ymax;
		cin >> x >> y >> r;
		xmin = max(x - r, 0);
		xmax = min(x + r, 10000);
		ymin = max(y - r, 0);
		ymax = min(y + r, 10000);

		for (int xi = xmin; xi <= xmax; ++xi) {
			Goblin finder(xi, ymin, 0);
			pair<set<Goblin>::iterator, bool> res = goblins.insert(finder);
			set<Goblin>::iterator it = res.first;
			set<Goblin>::iterator first;
			bool found = false;
			if (res.second) {
				goblins.erase(it++);
			}

			for (; it != goblins.end() && it->x == xi && it->y <= ymax; ++it) {
				if (in_range(*it, x, y, r)) {
					if (!found) {
						first = it;
					}
					found = true;
					count += it->count;
				} else if (found) {
					break;
				}
			}
			if (found) {
				goblins.erase(first, it);
			}

		}

	}

	cout << nbr_goblins - count << endl;
	return 0;
}

File: goblingardenguards/goblingardenguards.py
import math

class Goblin:

    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.count = 1

    def is_close(self, x, y, r):
        return math.sqrt((self.x - x) ** 2 + (self.y - y) ** 2) <= r

    def __str__(self):
        return str(self.x) + ", " + str(self.y) + ": " + str(self.count)

def range_compare(g, x, minY, maxY):
    g_cmp = g.x * 10001 + g.y
    minVal = x * 10001 + minY
    maxVal = x * 10001 + maxY
    if g_cmp < minVal:
        res = -1
    elif g_cmp > maxVal:
        res = 1
    else:
        res = 0
    return res

def find_between(goblins, start, end, x, minY, maxY):
    if start == end:
        return []
    while start < end:
        mid = start + (end - start) / 2
        g = goblins[mid]
        g_cmp = g.x * 10001 + g.y
        minVal = x * 10001 + minY
        maxVal = x * 10001 + maxY
        if g_cmp < minVal:
            comp = -1
        elif g_cmp > maxVal:
            comp = 1
        else:
            comp = 0
        if comp == 0:
            res = [ mid ]
            i = mid-1
            while i >= start and goblins[i].x == x and goblins[i].y >= minY:
                res.append(i)
                i -= 1
            i = mid+1
            while i < end and goblins[i].x == x and goblins[i].y <= maxY:
                res.insert(0, i)
                i += 1
            return res
        elif comp > 0:
            end = mid
        elif comp < 0:
            start = mid + 1
        else:
            return []
    return []

def main():
    nbr_goblins = int(raw_input())

    goblin_map = {}
    for i in range(nbr_goblins):
        x,y = [ int(val) for val in raw_input().split() ]
        if not (x,y) in goblin_map:
            goblin_map[(x,y)] = Goblin(x, y)
        else:
            goblin_map[(x,y)].count += 1

    goblins = list(goblin_map.itervalues())
    goblins.sort(key=lambda g: g.x * 10001 + g.y)

    count = 0

    nbr_spinklers = int(raw_input())
    for i in range(nbr_spinklers):
        x,y,r = [ int(val) for val in raw_input().split() ]

        minX = max(x - r, 0)
        maxX = min(x + r, 10000)
        minY = max(y - r, 0)
        maxY = min(y + r, 10000)

        for x_i in range(minX, maxX + 1):
            res = find_between(goblins, 0, len(goblins), x_i, minY, maxY)
            for i in res:
                g = goblins[i]
                if g.is_close(x, y, r):
                    count += g.count
                    del goblins[i]

    print nbr_goblins - count

if __name__ == "__main__":
    main()
